rescript

```diff
@@ -318,251 +197,382 @@ if (
+    var WebSocket = require("./lib/minisocket.js").MiniWebSocket;
+    var id = setInterval(notifyClients, 3000);
+    require("http")
+      .createServer()
+      .on("upgrade", function (req, socket, upgradeHead) {
+        dlog("connection opened");
+        var ws = new WebSocket(req, socket, upgradeHead);
+        socket.on("error", function (err) {
+          dlog(`Socket Error ${err}`);
+        });
+        wsClients.push(ws);
+      })
+      .on("error", function (err) {
+        // @ts-ignore
+        if (err !== undefined && err.code === "EADDRINUSE") {
+          var error = isTtyStd ? `\x1b[1;31mERROR:\x1b[0m` : `ERROR:`;
+          console.error(`${error} The websocket port number ${webSocketPort} is in use.
+Please pick a different one using the \`-ws [host:]port\` flag from bsb.`);
+        } else {
+          console.error(err);
+        }
+        process.exit(2);
+      })
+      .listen(webSocketPort, webSocketHost);
+  }
+
+  function watchBuild(watchConfig) {
+    var watchFiles = watchConfig.dirs;
+    watchGenerated = watchConfig.generated;
+    // close and remove all unused watchers
+    watchers = watchers.filter(function (watcher) {
+      if (watcher.dir === bsconfig) {
+        return true;
+      } else if (watchFiles.indexOf(watcher.dir) < 0) {
+        dlog(`${watcher.dir} is no longer watched`);
+        watcher.watcher.close();
+        return false;
+      } else {
+        return true;
       }
-      if (code !== 0) {
-        console.log(log);
+    });
+
+    // adding new watchers
+    for (var i = 0; i < watchFiles.length; ++i) {
+      var dir = watchFiles[i];
+      if (
+        !watchers.find(function (watcher) {
+          return watcher.dir === dir;
+        })
+      ) {
+        dlog(`watching dir ${dir} now`);
+        var watcher = fs.watch(dir, onChange);
+        watchers.push({ dir: dir, watcher: watcher });
       } else {
-        console.log(log, Math.floor(updateFinishTime() / 1e6), "mseconds");
+        // console.log(dir, 'already watched')
       }
     }
+  }
 
-    function logStart() {
-      updateStartTime();
-      let log = `>>>> Start compiling`;
-      if (std_is_tty) {
-        log = "\x1b[36m" + log + "\x1b[0m";
-      }
-      console.log(log);
+  /**
+   * @param {string | null} fileName
+   */
+  function checkIsRebuildReason(fileName) {
+    // Return true if filename is nil, filename is only provided on Linux, macOS, Windows, and AIX.
+    // On other systems, we just have to assume that any change is valid.
+    // This could cause problems if source builds (generating js files in the same directory) are supported.
+    if (!fileName) return true;
+
+    return !(
+      fileName === ".merlin" ||
+      fileName.endsWith(".js") ||
+      fileName.endsWith(".mjs") ||
+      fileName.endsWith(".cjs") ||
+      fileName.endsWith(genTypeFileExtension) ||
+      watchGenerated.indexOf(fileName) >= 0 ||
+      fileName.endsWith(".swp")
+    );
```

Comment:
I have a big question about the logic. Why not check only `.res`,`.resi`,`.ml`,`.mli` changes?
Also, since the whole condition is negated the `watchGenerated.indexOf(fileName) >= 0` doesn't trigger rebuild on generated file change 😅

Reply:
For example, if ReScript is used in a codebase mixed with ts. It'll recompile on every ts file change

Reply:
Good question. Maybe @bobzhang can comment?